service/history/timerQueueProcessor.go

```diff
@@ -717,6 +722,42 @@ Update_History_Loop:
 	return ErrMaxAttemptsExceeded
 }
 
+func (t *timerQueueProcessorImpl) processWorkflowTimeout(
+	context *workflowExecutionContext, task *persistence.TimerTaskInfo) error {
+	t.metricsClient.IncCounter(metrics.TimerTaskWorkflowTimeoutScope, metrics.TaskRequests)
+	sw := t.metricsClient.StartTimer(metrics.TimerTaskWorkflowTimeoutScope, metrics.TaskLatency)
+	defer sw.Stop()
+
+Update_History_Loop:
+	for attempt := 0; attempt < conditionalRetryCount; attempt++ {
+		msBuilder, err1 := context.loadWorkflowExecution()
+		if err1 != nil {
+			return err1
+		}
+
+		if !msBuilder.isWorkflowExecutionRunning() {
+			return nil
+		}
+
+		timeoutEvent := msBuilder.AddTimeoutWorkflowEvent()
+		if timeoutEvent == nil {
+			// Unable to add WorkflowTimeout event to history
+			return &workflow.InternalServiceError{Message: "Unable to add WorkflowTimeout event to history."}
+		}
+
+		// We apply the update to execution using optimistic concurrency.  If it fails due to a conflict than reload
+		// the history and try the operation again.
+		err := t.updateWorkflowExecution(context, msBuilder, false, nil, nil)
```

Comment:
We already have a helper method defined on the engine for updating history.  Why are you not using that?